services/brig/src/Brig/User/Auth.hs

```diff
@@ -78,78 +80,88 @@ data Access u = Access
       (return . userPhoneKey)
-      =<< lift (validatePhone phone)
-  user <- lift $ Data.lookupKey pk
+      =<< lift (wrapClient $ validatePhone phone)
+  user <- lift . wrapClient $ Data.lookupKey pk
   case user of
     Nothing -> throwE $ SendLoginInvalidPhone phone
     Just u -> do
       Log.debug $ field "user" (toByteString u) . field "action" (Log.val "User.sendLoginCode")
-      pw <- lift $ Data.lookupPassword u
+      pw <- lift . wrapClient $ Data.lookupPassword u
       unless (isNothing pw || force) $
         throwE SendLoginPasswordExists
       lift $ do
-        l <- Data.lookupLocale u
-        c <- Data.createLoginCode u
+        l <- wrapClient $ Data.lookupLocale u
+        c <- wrapClient $ Data.createLoginCode u
         void . forPhoneKey pk $ \ph ->
           if call
-            then sendLoginCall ph (pendingLoginCode c) l
-            else sendLoginSms ph (pendingLoginCode c) l
+            then wrapClient $ sendLoginCall ph (pendingLoginCode c) l
+            else wrapClient $ sendLoginSms ph (pendingLoginCode c) l
         return c
 
-lookupLoginCode :: Phone -> (AppIO r) (Maybe PendingLoginCode)
+lookupLoginCode :: (MonadClient m, Log.MonadLogger m, MonadReader Env m) => Phone -> m (Maybe PendingLoginCode)
 lookupLoginCode phone =
   Data.lookupKey (userPhoneKey phone) >>= \case
     Nothing -> return Nothing
     Just u -> do
       Log.debug $ field "user" (toByteString u) . field "action" (Log.val "User.lookupLoginCode")
       Data.lookupLoginCode u
 
-login :: Login -> CookieType -> ExceptT LoginError (AppIO r) (Access ZAuth.User)
+login ::
+  Login ->
+  CookieType ->
+  ExceptT LoginError (AppIO r) (Access ZAuth.User)
 login (PasswordLogin li pw label _) typ = do
   case TeamFeatureSndFPasswordChallengeNotImplemented of
     -- mark this place to implement handling verification codes later
     -- (for now just ignore them unconditionally.)
     _ -> pure ()
-  uid <- resolveLoginId li
+  uid <- mapExceptT wrapClient $ resolveLoginId li
   Log.debug $ field "user" (toByteString uid) . field "action" (Log.val "User.login")
-  checkRetryLimit uid
-  Data.authenticate uid pw `catchE` \case
-    AuthInvalidUser -> loginFailed uid
-    AuthInvalidCredentials -> loginFailed uid
+  mLimitFailedLogins <- view (settings . to Opt.setLimitFailedLogins)
+  mapExceptT wrapClient $ checkRetryLimit mLimitFailedLogins uid
+  mapExceptT wrapClient (Data.authenticate uid pw) `catchE` \case
+    AuthInvalidUser -> mapExceptT wrapClient $ loginFailed mLimitFailedLogins uid
+    AuthInvalidCredentials -> mapExceptT wrapClient $ loginFailed mLimitFailedLogins uid
     AuthSuspended -> throwE LoginSuspended
     AuthEphemeral -> throwE LoginEphemeral
     AuthPendingInvitation -> throwE LoginPendingActivation
   newAccess @ZAuth.User @ZAuth.Access uid typ label
 login (SmsLogin phone code label) typ = do
-  uid <- resolveLoginId (LoginByPhone phone)
+  uid <- mapExceptT wrapClient $ resolveLoginId (LoginByPhone phone)
```

Comment:
Maybe we should add a helper function `wrapClientE = mapExceptT wrapClient`?

Reply:
Thanks! I just added it in an upcoming commit.